# 프로그램에서의 에러 종류

프로그램 실행 중에 메모리 등 컴퓨팅 자원 부족이나 사용자의 잘못된 입력 등 여러 원인으로 프로그램이 비정상적으로 종료하는 
경우가 있다. 이는 프로그램이 문법적으로 작성이 잘못된 것이 아니라 프로그램의 외적인 요인으로 발생하는 것으로 이러한 예외를 
올바르게 처리하지 못해 오류로 이이진다면 이는 프로그래머의 잘못이기 때문에 예외 상황을 적절하게 대처하는 프로그램을 작성하는 것이 필요하다. 

## Compile errors (컴파일 에러)
컴퓨터 프로그램에서 컴파일 에러는 구문 에러(syntax error)라고도 하며 작성된 프로그램 코드가 프로그램 언어의 문법에 위배되는 경우에 발생한다. 

## Logic errors (논리 에러)
컴퓨터 프로그래밍에서 논리 에러는 프로그램이 비정상적으로 종료되지 않으나 올바르지 않게 동작되는 프로그램의 버그로 의도되지 않거나 원하지 않는 결과 또는 동작이 발생한다. 
아래의 코드는 평균을 계산하려고 하는데 결과는 a의 값에 b/2를 더한 연산이 된다. 
```
double average (double a, double b) {
  return a + b / 2;      // (a + b)/2
}
```

아래의 코드는 a 가 0 인 조건이 만족될 떄 아래의 명령문을 실행하고자 하나 a가 0인 경우에도 조건문 내의 문장이
실행되지 않는 문제가 발생한다.

```
if (a = 0) {   //  a == 0
	....
}
```
## Runtime errors (런타임 에러)
컴퓨터 프로그래밍에서 런타임 에러는 프로그램 실행 중에 발생하는 프로그램의 버그로 프로그램 중 0으로 나누기 에러, 동적 메모리 할당 시 메모리 부족 에러, 배열에서 범위를 벗어난 영역의 접근 에러가 예이며 컴파일 과정에서는 검출되지 않는다.
이 에러는 프로그램이 종료되거나 잘못된 결과가 나온다.
아래의 프로그램은 정수의 팩토리얼 값을 구하는 함수를 재귀 함수의 형태로 정의한 후 ```main```함수에서 3의 팩토리얼 값을 구하는 함수를 호출한다.
이 프로그램을 실행시키면 ``segmentation fault``가 발생하고 프로그램이 종료된다. 
이는 ```factorial```함수 내에서 ```factorial```함수를 호출하는데 
```factorial```함수 내에서 호출을 중단하는 조건이 없어서 함수를 계속 호출하게 된다.
C++ 프로그램에서 함수를 호출할 때 마다 함수 호출과 관련한 데이터를 스택에 저장한다.
함수가 무제한 호출됨에 따라 스택의 저장공간을 모두 소진하게 되어 런타임 에러가 발생하게 된다.   

```
#include <iostream>
using namespace std;

int factorial(int n) {
	return n * factorial(n-1);
}

int main(int argc, char const *argv[])
{
	cout << factorial(3) << endl;
	return 0;
}
```


